### ISSUE TITLE
s/`parse_backslash`/`strict_backslash_esc`/ and show retained and gained compatibility

### ISSUE THREAD

**Issue by bar-g** [oilshell/oil#1818]
The conclusion after the discussion of the original problem in this issue was that the `parse_backslash` implementation is fine (i.e. no verbatim or quoting backslash).

[Because what it does is printing errors on escape sequences not supported in the stricter ysh. The adapted code (supported subset) remains 100% compatible with shell, and ~~double-quotes in python and others are fully compatible with it as well~~partly others.]

However, the option should be renamed to `strict_backslash_escape` (i.e. all backslashes are escape characters, no verbatim or quoting backslash), and `shopt -p` could also print its motivation: How complying code is compatible within shell and many others:
```
ysh$ shopt -p | grep strict_backslash_escape_err
shopt -s strict_backslash_escape   # no verbatim or quoting backslash (compat: =>shell->adapt, =>many->adapt)
```
Meaning:
`=>shell->adapt `
* Affected code executing with this option is compatible shell code.
* Shell code will need adaption to execute with this option (i.e. remove faux escapes, and double-up printed backslashes).
* A longer thin arrow (`-->`) would mean it's not just a limited/small (mutating) adaption.

----

[Orig. problem and an alternative but less compatible solution as devised during the discussion below:]

**Problem:**

1. There should only be one way to write a string (per quotation-type), not different escaped variants that print the same.

2. It should work without breaking all regular (verbatim) backslash usage, i.e. forcing addition of an escape character for each backslash to print.
   ```
    ysh ysh-0.19.0$ echo "c:\file"
   
     echo "c:\file"
             ^
   [ interactive ]:727: Invalid char escape in double quoted string
   ```
3. Instead of also breaking repeated backslashes as with the old `parse_backslashes`, which requires to minutely count backslashes, and shows alternating error- instead of shells character-dropping behavior, the verbatim text expression capabilities should be improved, instead.

**Solution:**

A new `strict_backslash_escape` that does not break regular backslash usage. (Only interprets backslashes if located strictly before some special escapable character.)

So, with the new `strict_backslash_escape` enabled, the `\` character continues to be usable just normally in all string quotation-types, except only...

...before characters that need to be escaped to be printed, that is e.g. just:

* **before escapable characters**, e.g. `"` and `$`, where it functions as an escape character, unless
* **repeated twice before escapable characters** (`\\$` and `\\"`) to make it a "written" escape character, and
* if **repeated three times before the quoting character** (e.g. `\\\"`), to make it a verbatim backslash before an ending quotation mark.

It would actually work mostly the same as before, but with the ambiguity removed and improved verbatim printing of backslashes, even if repeated:

**Table:**
(Updated from orig. version in https://github.com/oilshell/oil/issues/860)

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;code&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; | shell | first implementation: "breaking" `parse_backslash` (Oil)  | new: "compat" `strict_backslash_escape` w/ improved verbatim capabilities (YSH) 
---------------- | -- | -- | --
`echo \n` | literal `n` ! | ERROR | error (asking for removal or quotation, to get shell-compat)
`echo \\` | literal `\` | compat | error (asking for quotation, to get shell-compat)
`echo "\n"` | literal `\n` | ERROR (breaks copy-pasted texts, and verbatim typing) | compat (backslashes as typed or copy-pasted)
`echo "\\" `| literal `\` (mangling) | compat (mangling) | compat (mangling, but only due to double-backslash-before-quote)
`echo "\\ "` | literal `\ ` (mangling) | compat (mangling) |VERBATIM literal `\\ ` (as typed, more copy-pasted texts work as is)
`echo "\\\ "` | literal `\\ ` (mangling, if chaining odd amount) | ERROR if chaining odd amount |VERBATIM literal `\\\ ` (any chained amount)
`echo '\n'` | literal `\n` | compat | compat
`echo  '\\'` | literal `\\` (no backslash escapes) | compat | compat
`echo $'\n'` | newline | compat | compat
`echo $'\\n'` | literal `\n` | compat | compat (but due to double-backslash-before-escapable)
`echo  $'\\'` | literal `\` (mangling) | compat | compat (but due to double-backslash-before-quote)
`echo  $'\\ '` | literal `\ ` (mangling) | compat |VERBATIM literal `\\ `
`echo  $'\\\ '` | literal `\\ ` (mangling) | ERROR if chaining odd amount |VERBATIM literal `\\\ ` (can chain any amount)
`echo  $'\$'` | literal `\$` | ERROR (should be $'\\\\$') | compat

**Examples:**

```
$ echo " \  \z \\z"  # regular backslashes
 \  \z \\z
$ echo " \" \$ "      # characters needing escapes
 " $ 
$ echo " \\" \\$ "     # chars needing escapes with escape char written before them
 \" \$ 
$ echo "  \\\"           # escape char written before a quotation-ending quotation mark
  \
$ echo " \\\" \\\$ \\\\"   # all further "prefix" escape characters print normally (after the "escape offset")
 \\" \\$ \\
```

There is a small behavioral difference in the escape-char-before-quote-ends edge case, which however allows for better error behavior than before, and can present a helpful error message:

```
# ordinary shell
osh $ echo "  \" # backslash at end of string
>                        #=> unclosed quotes causing a second level line prompt, or script breakage
osh $ echo "  \\"
  \
```
```
# new ysh
ysh $ echo "  \" and also "  \\"  #=> Error message: Unclosed double-quote (use \\\" to close after backslash).
ysh $ echo "  \\\"
  \
```

----

[Old originally filed issue, that was still mixing up different things:]

I was using the following to print two closing characters, i.e. `'\n` (a single-quote plus a newline), and it worked:

```
~ $ printf  "'\n"
'
~ $
```

However, in ysh this produces an error:

```
ysh ysh-0.19.0$ printf "'\n"

  printf "'\n"
           ^
[ interactive ]:633: Invalid char escape in double quoted string

```

1. Did I miss this breakage in https://www.oilshell.org/release/0.19.0/doc/upgrade-breakage.html , or is it actually missing?

2. The error might be good if it mentions to use `$'...'` C-style backslash escaped strings, **if run with some strict_ option for ysh:upgrde**, however, shouldn't it just print `'\n` (verbatim, no newline character) in the case at hand here, i.e. with ysh:all (leaving the case to be handled by some follow up error -- solvable with printf '%b' backslash interpretation -- if a newline character is really wanted)?

**Comment by bar-g** [comment]
Hm, `printf` is supposed to interpret the format string itself.

But `echo` actually does seem to print it only verbatim in ordinary shells:


```
~ $ echo "'\n"
'\n
~ $ 

```

```
ysh ysh-0.19.0$ echo "'\n"

  echo "'\n"
         ^
[ interactive ]:674: Invalid char escape in double quoted string
```

So, at the moment I don't know, why is it invalid?

**Comment by andychu** [comment]
Hm can you post the whole code snippet?  We have an idea to fold `printf` into double quoted strings, like `"number = ${x %2f}"`, which will make it obsolete


The "correct" thing is to do

```
printf  "'\\n"
```

Because `\\` is a backslash inside double quotes, and then `printf` is another dynamically parsed language

We want to move away from both `printf` and `echo -e` because they're dynamic, not static.  They are separate languages from shell


----

So this error is intended but there should be a nicer way to write the code

**Comment by bar-g** [comment]
To give some context, what I'm currently testing (and filing faced issues about) is combining "fully-featured" ysh code and compatible shell code. I'm impressed how well it's going, and find it really useful, e.g. to include same shell code on non-ysh router and on ysh desktop.

---

I kind of have an answer, but if I could better understand your reason/intention for breaking double-quotes, as it seems, it could become easier/clearer or even much revised.  Could you please add some reasoning?

(I mean concerning the "Invalid char escape in double quoted string". I guess printf is a separate thing.)

**Comment by bar-g** [comment]
Which shopt is this actually?

**Comment by andychu** [comment]
It should be `parse_backslash`, so yeah you can simply turn that on you don't want it

```
$ bin/ysh  -c 'echo "\n"'
  echo "\n"
        ^
[ -c flag ]:1: Invalid char escape in double quoted string
```


```
$ bin/ysh -o parse_backslash -c 'echo "\n"'
\n
```

I thought I called it `strict_backslash`, but it's `parse_backslash` I guess because it's a parsing option.  And it's **on** in OSH but **off** in YSH, so you have to turn it back on.

`shopt -p` prints the current option settings




-----------


The best list of options is here (I will update this topic slightly)

https://www.oilshell.org/release/0.19.0/doc/ref/toc-ysh.html#option

**Comment by andychu** [comment]
The basic reason is that to be very clear about what code means

This is confusing, there should only be one way to write the string (the first way)

```
$ bin/osh -c 'echo "\\z"'
\z

$ bin/osh -c 'echo "\z"'
\z
```

Or use `echo '\z'`

There are too many ways to quote things in shell, so we reduce them

**Comment by andychu** [comment]
I wonder if it's less confusing to say `parse_strict_backslash` and `parse_strict_backticks`

Maybe too late for that though

**Comment by bar-g** [comment]
Thanks! The background is helping a lot.

First, the generic thing.

Concerning confusing shopt names, I don't think is too late. It's always good to clear up confusion from previous iterations, even if there were thousands or much much more users already "add-deprecate-(keep/warn/remove)". It'll help the understanding for them and foremost for each and every one still to come.

I came to realize good strict_* and parse_* option patterns between the elevated error/warning feedback "in the middle" of osh/ysh and both sides, can clear up a lot.
* compatible-osh
* strict:all-osh (many errors or warnings)
* ysh:upgrade (some errors and hints)
* ysh:all

See the case around `echo`, it has shown that confusion can be much reduced by having corresponding strict_* and parse_* options, to control error/warnings and behavior separately (https://github.com/oilshell/oil/issues/1772).

----

Ok, so there should only be one way to write a string.

But does that really require to dump and forbid all regular backslash usage?

Couldn't the `\` character be used normally in double-quoted strings with a `strict_backslash_escape`, except...
* before the expansion characters `"` and `$`, where it functions as an escape character, unless
* if repeated twice before the expansion characters (`\\$` and `\\"`) to make it a "written" escape character, and
* if repeated three times before a double quote, to make it a verbatim backslash not escaping the end of the quotation.

```
$ echo "\\z"
\\z
$ echo "\z"
\z
$ echo " \" \$ "     # escapes
 " $ 
$ echo " \\ \\" \\\"   # written escapes
 \\ \" \
$ echo " \\\ \\\" \\\\"   # normal additions after the "escape offset"
 \\\ \\" \\
```

Puh, a `strict_backslash_escape` may not completely break backslashes in double-quotes if it could work like that.
Actually working mostly the same, with only some differences in edge cases.

```
osh $ echo "\"               #=> error, unclosed
osh $ echo "\\"
\
ysh $ echo "\" and also "\\"  #=> error, unclosed (use \\\" to close after backslash)
ysh $ echo "\\\"
\
```

**Comment by andychu** [comment]
Those rules are too complicated!

And sorry I meant there is only 

- one way to write the string with double quotes
- one way to write it with single quotes

Not 2 ways with double quotes.

---

I think the real problem is that we need to fold `printf` into `""`


So that you never need to use printf in YSH

You should use either OSH and printf, or YSH and new style

---

I appreciate the transition is not as easy as it could be -- though I also am wary of adding a whole intermediate state

Basically I think if you start out with

```
printf 'hi %2f\n' "$x"
```

then in YSH you just change it to

```
echo "hi ${x %2f}"   # not implemented yet
```

That's not a trivial jump but the end result is cleaner

It would be nice if there was a way to auto-rewrite it, but that seems beyond the scope of the project now.  We are going to export the Oils lossless syntax tree (AST) -- so other people could do this, as tools of their own

**Comment by bar-g** [comment]
Maybe take a bit more time to think it through.

I understand that in the parser the proposed rules may be a bit more complicated to implement.

However when writing a double-quoted string text,  the adding of the escape character only at the places where needed, and it having a special meaning only at those places (and none elsewhere), is what really works nicely and naturally.

Compare that to having to escape every backslash (`\\`) when writing a double-quoted string to avoid ysh errors, and multiple backslashes in a row silently dropping out (shell) or producing errors on every other odd.

Consider the minimized difficulties for copy pasting arbitrary text into a double-quoted string!

---

I don't understand, you mean one same way to write double- and single-quoted strings?

I think the `strict_backslash_escape` I propose would produce mostly the same way in ysh as in osh. However, because the escape character is only an escape character at places where there actually is something to escape, all ambiguity of nonsensical character escaping is removed.

So, YSH's double-quotes would be more compatible with generic text. Ordinary shells require fiddling with multiple backslashes all over the place, not just for `$`, `"`, and `"` at the end. 

---

I think "printf-style interpolation" would be another really nice ysh feature. But it was just unfortunate that I stumbled upon the broken double-quotes in ysh while using printf. (Already changed the issue title to using `true`.)

Breaking regular backslashes in double-quotes has much wider effects: string variable assignments, specifying command arguments, ... .

----

> You should use either OSH and printf, or YSH and new style

I'm using shell scripts _**and**_ YSH scripts, calling and sourcing each other. And I'm sure there are many sites that will also continue to be working with shell scripts for some time to come.

And being able to use the same double-quoted string format (except edge cases) in both areas (compatible shell and ysh) seems quite crucial.

How do you picture pasting strings and texts to $interpolate between scripts, and from outside material?

**Comment by andychu** [comment]
I don't think sourcing OSH and YSH from each other is a good idea

I would limit it to invoking them as subprocesses.   e.g. `./foo.ysh` rather than `source foo.ysh`

Then you get a fresh set of options each time

...

I don't see the problem, because you can simply double up the escapes.

If you really want to use double quoted strings, then you write `"\\n"` and not `"\n"`.  I wouldn't write it that way, but it works.

What's the problem?

**Comment by bar-g** [comment]
Hi, yeah, I think whether OFF->ON is good depends on the option name, though.

If its as clear like `echo_flags` it seems completely natural to turn it off for ysh. While it seems natural that `strict_*` options get turned on.

For the `parse_*` options maybe it could work out if you remove the parse_ prefix, and maybe add some more context instead?

Kind of shows the problem of this issue: A general `shopt --unset backslashes` makes it obvious that this option is disabling too much!

---

> I don't see the problem, because you can simply double up the escapes.

Nah, that is a problem. That's a tedious unnecessary chore and breaks compatibility where there is no need to.

---


See, I found the original table from when you implemented `parse_backslash`, and I've now spend some time on working this out a bit better. Please open the updated issue description, it now even has a table comparison, and much better structured examples.

**Comment by andychu** [comment]
Yeah I looked at the table, it's changing the meaning of existing shell code too much.

What we are doing in YSH is (optionally) **removing** some ways of expressing the same thing, because it's too much freedom -- shell is too noisy and accepts too much


According to our  https://github.com/oilshell/oil/wiki/Language-Design-Principles  we want to avoid "taking over" existing syntax

(The two major exceptions being word splitting "silently" changing, and  `echo -e` silently changing -- I view those as fixing existing bugs.)

But here we are not changing shell, so much as taming the possibilities, making it easier to read code.

Users aren't going to be able to remember what's in your table -- it has to be smaller than that.

---

Also the proposal WIDELY diverges from what `\` already means in every language.  Python and JavaScript both use the same rules for `\`. 

In Python and JS, `'\\'`  is a single backslash.  And in C.  It's very well accepted



----

BTW `$'\\'` is now legacy -- in Oils 0.20.0 we will have "J8 strings" which look like `u'\\'` and `b'\\'`.    They  can express arbitrary byte strings and nice unicode escapes.

**Comment by bar-g** [comment]
Hm, ok, how suboptimal can changing (actually improving verbatim writing) the existing syntax be here?

Do you have an example, where it's bad to allow singleton backslashes (within quotes, not command language), and where it's better to forbid and require "escaped escapes"?

I mean, why shouldn't users just re-enable `parse_backslash` right away?


> Users aren't going to be able to remember what's in your table -- it has to be smaller than that.

Hm, the only special things about backslashes to remember would be the three items written in bold (escapes are escapes only before escapable chars, repeat twice (only there) to write backslash, and tripple it before a quotation mark to still get a valid ending-quote).

> Python and JS [...]. And in C. It's very well accepted

Uh, unfortunate early design choice. Well, shell is for succinct interactive use, so probably that's why they agreed to allow singleton backslashes within quotes (and also allowed in command language but there they get lost).

---

At least I think this might have found a better name for `parse_backslash`, maybe `singleton_backslashes` (disabled in ysh)? Or, something denoting whatever the risk of re-enabling it is.

And a better error message:
`Invalid singleton backslash (char escape) in double quoted string (use \\ to express one).`

**Comment by bar-g** [comment]
Ok, singleton backslashes (when lost or same as double-backslash) are bad for comparing strings. Anything else?

And a changed (instead of a subset) format within double-quotes does not allow collective shared usage of variables from ysh and osh. So, nah, I actually don't want that incompatibility at all, I want compat as much as possible.

---

How come you also couldn't give these answers say at the beginning? Would have guessed it's much easier for you to quickly see the conclusions, than for me. :-)   [In any case, my thanks, I much appreciate your help.]

**Comment by bar-g** [comment]
Hm, what about calling it `break_backslash`? (That could actually apply to all parse_* options of the breaking group.)

But it could be even better: More self explaining, when naming those according to the pitfall, fault, or thing, they prevent/or fix.

So, it again depends on the question: What is `break_backslash` good for? (Or why make it default or not?)

Is it `break_manifold_strings`  or `enforce_unique_string_enc`?

---

> sourcing OSH and YSH from each other

I think I want it because I want to share state between ysh modules and some remaining osh modules. Passing data isn't nice, because it's not simple to use json in osh. But accessing string variables is, and it seems to works nicely when only using string variables in the ysh parts!

But for sure, it would only be complete if osh can also access (global) ysh types (of course only at the basic string level interface).

This would allow the ysh parts to make full use of the language, while osh parts still have a defined string interface to the data, i.e. https://github.com/oilshell/oil/issues/1813 "ysh-list/dict not accessible with osh syntax (sourcing ysh procs)" and https://github.com/oilshell/oil/issues/1811 "fatal: "declare -n" refs to ysh vars (as universal interface "narrow waist" to source/use ysh code)".

---

> you write "\\n" and not "\n". I wouldn't write it that way, but it works.

What would you write instead? That could be helpful and shortening.

**Comment by bar-g** [comment]
Urgh, this ugly backslash mangling is even present in these github issue comments! It's awful.

See how the quote in my reply displays, it becomes  `"\n" and not "\n"`, even though you originally wrote two backslashes on the left, and they are there in the Markdown comment source.

**Comment by bar-g** [comment]
> I wonder if it's less confusing to say parse_strict_backslash and parse_strict_backticks

Hm, counting on tab-completion, it shouldn't be a problem if options get slightly longer.  The least confusing may be names that most precisely explain their effects.

The current `parse_backslash` actually is only breaking in one direction, is this correct?

I think the shopts naming in general could remove much confusion by adding some classification prefixes to the names, which could indicate backwards/forwards code compatibility (c) or incompatibility (i). 
Also may be indicate smaller breakages (i) vs. large breakages (I).

What do you think?:

I think the individual option switches are really mostly only for interested or "expert" users, anyway, so I think it's good to provide that info, and fine to use some kind of "cryptic" prefix, that may perfectly be ignored otherwise.

So, for the current `parse_backslash`, how about this:

It re'strict's the expression of backslashes, only allowing their escaped expression. Existing shell _code_ becomes incompatible, because the option breaks all singleton backslash usages at all places (the ysh _interpreter_ is not "backwards compatible" with this option), however ysh _code_ running with it remains compatible shell code.

So referring to the code, this option introduces a larger incompatibility for shell code to run in ysh, but adapted (ysh) code remains compatible to be used in shell.

From this I'd derive a name something like this:

`Ic_strict_escaped_backslash_only`
With it's error messages printing a helpful `Ic_strict_escaped_backslash_only: backslashes can only be expressed escaped ('\\')`.

Applying the same logic to what I proposed above as `strict_backslash_escape`, actually had the nice effect to allow devising an option with the same characteristics but with much improved compatibilty!:

---

As you explained, looking at what I proposed before as a separate option, it would have been a new/changing feature, that would break less verbatim string expression than the strict restriction to only escaped backslashes, however creating limited incompatibilities in _both_ directions, i.e. prefixed 'ii_':

 `ii_only_char_bound_escapes` or, put positively `ii_unbound_non_escaping_backslash`
This is clearly indicating a breaking feature. Even though the incompatibility would really only concern chained backslashes therein (these not being "doubled up").


So now for a combination with much better shell compatibility (and interactive typing usability):

A low-impact strict_ option that guarantees unique string encoding while maintaining full compatibility for single backslashes in both directions, for both the escaping ones and non-escaping verbatim ones.

The "trick" is to forbid singleton backslashes and odd-length-chains in command language (just at now), but in quoted strings, forbid single-pairs and odd-length-chains, but not singleton backslashes.

May this actually maintain full string-compatibility for single escaping backslashes, as well as for single non-escaping verbatim backslashes? Only erroring on superfluous backslashes, is that what you call "bad backslashes"? So could it be named a nice two-way compatible strict option?

`cc_strict_unique_string_encoding`

It's nothing new if single backslashes behave differently in command language than within quoted strings.

It could better fulfill the principle to maintain compatibility on the top-level, i.e. basic command language and quoted strings keep working in the same shell way. The strict option only forbids lesser used, redundant things (singleton-\ in commands, singleton-\\ in quoted strings).

It's also compatible with the python/js etc. way with only backslash escapes or backslash pairs.

**Comment by bar-g** [comment]
An even better name for `parse_backslash` seemed `Ic_strict_escaping_backslash`, so I changed that above.

Now, I even realize all my confusion with python/js etc. compatibility could have been avoided from the beginning, say by a compatibility suffix:


`Ic_strict_escaping_backslash_cc`

So the larger incopatibility on the (left) for existing shell code is warranted by full both-way compatibility on the (right) python/js side!

**Comment by bar-g** [comment]
Likely better to read, sort, and use:

Only have `shopt -p` print the full info up front, but only as an optional part, that's not required for manually switching the option:
`Ic-cc:strict_escaping_backslash`?

**Comment by bar-g** [comment]
I am surprised how much relieve and understanding these compatibility info bits give.

I think providing such compatibility and incompatibility info is a crucial part to foster understanding and appreciation of what comprises and constitutes osh and ysh.

Maybe still to find a better representation for the `shopt -p` output?:
```
shopt -s strict_escaping_backslash   #  =>shell->adapt ,  =>python=>compat
```

**Comment by bar-g** [comment]
Meaning:
`=>shell->adapt `
* Affected code executing with this option is directly compatible with shell code.
* Shell code will need adaption to execute with this option

Example of incompatible, "added features" option:
`adapt->shell=>compat`
* Affected code executing with this option may need adaption to run in shell.
* Shell code is compatible to execute with this option.

**Comment by bar-g** [comment]
Thanks here as well,
to finish it up, this is the conclusion I draw from this issue's discussion:

---

The conclusion after the discussion of the original problem in this issue was that the implementation of the current `parse_backslash` is fine, because adapted code remains 100% compatible with shell, and double-quotes in  python and others are fully compatible with it as well.

However, better warrant the required adaptions, by making the retained and gained compatibility much more obvious: 

The option (as-it-is) should be renamed to `strict_backslash_escape`, and `shopt -p` could show how the complying code is directly compatible with shell and python:
```
ysh$ shopt -p | grep strict_backslash_escape
shopt -s strict_backslash_escape   # if set:  =>shell-->adapt ,  =>python=>compat
```
Meaning:
`=>shell-->adapt `
* Affected code executing with this option is compatible shell code.
* Shell code will need adaption to execute with this option (i.e. remove faux escapes, and double-up printed backslashes).
* The longer thin arrow means it's not just a limited/small (mutating) adaption.

`=>python=>compat`
* Affected code executing with this option is compatible python code.
* Python code is compatible to execute with this option.

**Comment by bar-g** [comment]
Actually, ysh does not seem to be compatible with python at all.

What are you telling!

```
$ python3
Python 3.11.2 (main, Mar 13 2023, 12:18:29) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> print("m \n ")
m 

>>> 
>>> print("m \\ ")
m \ 
>>> print ('m \ ')
m \ 
>>> print ('m \" ')
m " 
```

**Comment by andychu** [comment]
Well I wasn't saying that YSH behaves exactly like Python, but I was saying that `\\` means a single backslash in Python, 
JavaScript, and C

I think some of your proposed semantics made `\\` into 2 backslashes, I don't remember though

---

What I will say is that there is a common subset of YSH and Python that is very close

**Comment by andychu** [comment]
One advantage of the strict behavior is that most people can't remember which language supports what escapes.

Python supports `\0`

```
$ python -c 'print("\0")' | od -c
0000000  \0  \n
0000002
```

But shell doesn't

```
$ bash -c 'echo \0'
0
```

YSH notifies you that this is an error:

```
$ bin/ysh -c 'echo \0'
  echo \0
       ^~
[ -c flag ]:1: Invalid char escape (parse_backslash)

```

YSH is like JSON -- JSON disallows `\z`, but Python apparently doesn't

Most C compilers warn about bad escape sequences these days

**Comment by bar-g** [comment]
Thanks for explaining! Now I'm sure I just had misunderstood you before. (Possibly kind of some wishfully thinking on my side that general progaming languages could have been more consistent.)

Your explanations actually also seem to be a nice description for, and possibly also a good reason to still rather rename this option to something like `strict_backslash_escape_err`.

Printing errors on escape sequences that are not supported in shell is a good reason for forbidding singleton backslashes.

I adapted the option's description proposed in the issue description to this:

```
ysh$ shopt -p | grep strict_backslash_escape_err
shopt -s strict_backslash_escape_err   # not print unknown escapes (compat: =>shell-->adapt, =>several-->adapt)
```

**Comment by bar-g** [comment]
> (The two major exceptions being word splitting "silently" changing, and `echo -e` silently changing -- I view those as fixing existing bugs.)

I haven't thought about if it's also possible for word splitting, but I think the `echo -e` bug fix can be done nicely with good error messages, so not silently at all.

That is with `strict_echostrings` enabled in `ysh:upgrade` and disabled in `ysh:all` together with **_a separate_** `echo_flags` option that also gets disabled in `ysh:all` (#1772).

**Comment by andychu** [comment]
It might be reasonable to rename it to `strict_backslash` , because that will "invert" OSH->YSH from  `yes -> no` to `no -> yes`

Still thinking about making all the options turn ON by default for YSH

**Comment by bar-g** [comment]
Concerning turning off `echo_flags` I can't think of a better wording or understanding with turning an option on. 

It's so short, nice, and self explaining, or do you have an idea that fits into the explaining and setting better than disabling `echo_flags` in ysh?

**Comment by andychu** [comment]
Well it could be `shopt --set no_echo_flags`, or I honestly I think `shopt --set simple_echo` is an OK name

**Comment by bar-g** [comment]
Hm, simple_echo is kind of an OK name, only if one already knows what it does. But without knowing, newcomers (my experience) first have to ask what is simple? (And then go, why that? If they used -en before-)

I think `no_echo_flags` is more directly understandable if you really want strict (or pedantic?) "all options 'on' in ysh". Because, it says what it does. (However, it adds a different, avoidable confusion though, because you need to --set something (on) to turn it off.)

I mean all-on works nice for `strict_*` options, but naturally some legacy `parse_*` options need to be turned off and newer ones turned on (feature flags), also there could be some common standard (bash compat?) feature names that are turned off and should not get renamed into a `no_*` option. So, all in all it seems, a telling name that can be intuitively --set and --unset is worth something in itself.

To still help the systematic cause, maybe there could be something like shopt --print mentioning the ysh default in a comment? And shopt --print --ysh could print setting all options to ysh defaults for comparisons?

---

(Just noticed, looking through shopt -p, wouldn't it make sense for `dynamic_scope` to be `_dynamic_scope`? Isn't it usually just auto-off within funcs/procs? And on, on top-level and in shell functions.)

---

Finally, back to the problem of the why?-question, that again can be solved nicely by providing an error message that readily contains corresponding help-on-error hints when the users hits on it.

See the error message example that I added at:
*  https://github.com/oilshell/oil/issues/1772
